File: src/cirs_girona_cala_viuda/scripts/frontend/front_end_isam.py
import gtsam
from functools import partial
import matplotlib.pyplot as plt
import numpy as np
from gtsam import NavState, Point3, Pose3, Rot3
from gtsam.symbol_shorthand import B, V, X
from mpl_toolkits.mplot3d import Axes3D
from scipy.spatial.transform import Rotation as R
from typing import Optional, List

from dataloader import *


class AUViSAM:
    def __init__(self):
        """
        The nodes on the graph will be gtsam.NavState, which is essentially a
        SE_2(3) lie group representation of the state of the vehicle.

        For this script, will be testing out the use of the IMU, depth sensor,
        odometry, and velocity logger.
        """

        # Initialization of some parameters
        self.dt = 1e-6
        self.priorNoise = gtsam.noiseModel.Isotropic.Sigma(6, 0.1)
        self.velNoise = gtsam.noiseModel.Isotropic.Sigma(3, 0.1)
        self.depth_model = gtsam.noiseModel.Isotropic.Sigma(1, 0.1)
        self.dvl_model = gtsam.noiseModel.Isotropic.Sigma(3, 0.1)

        # IMU shiz
        acc_bias = np.array([0.067, 0.115, 0.320])
        gyro_bias = np.array([0.067, 0.115, 0.320])
        bias = gtsam.imuBias.ConstantBias(acc_bias, gyro_bias)

        self.preintegration_parameters()

        self.pim = gtsam.PreintegratedImuMeasurements(self.params, bias)

        # Load data
        self.iekf_states = read_iekf_states("states.csv")
        self.iekf_times = read_state_times("state_times.csv")
        self.imu_times, self.imu = read_imu("full_dataset/imu_adis.csv")
        self.depth_times, self.depth = read_depth_sensor(
            "full_dataset/depth_sensor.csv"
        )

    # Setters

    def preintegration_parameters(self):
        """
        Set the parameters for the preintegration
        """
        self.params = gtsam.PreintegrationParams.MakeSharedU(9.81)
        I = np.eye(3)
        self.params.setAccelerometerCovariance(I * 0.1)
        self.params.setGyroscopeCovariance(I * 0.1)
        self.params.setIntegrationCovariance(I * 0.1)
        self.params.setOmegaCoriolis(np.zeros(3))

        self.bias_cov = gtsam.noiseModel.Isotropic.Variance(6, 0.1)

        self.delta = Pose3(Rot3.Rodrigues(0, 0, 0), Point3(0.05, -0.10, 0.20))

    # Getters

    def get_nav_state(self, time):
        """
        Get the state from the Invariant EKF at time "time" and store
        as a gtsam.NavState to initialize values and/or set nodes in the graph

        Inputs
        =======
        time: int
            Index of the time in the time vector

        Returns
        =======
        nav_state: gtsam.NavState
            The state at time "time"
        """
        x = self.iekf_states["x"][time]
        y = self.iekf_states["y"][time]
        z = self.iekf_states["z"][time]
        u = self.iekf_states["u"][time]
        v = self.iekf_states["v"][time]
        r = self.iekf_states["r"][time]
        phi = self.iekf_states["phi"][time]
        theta = self.iekf_states["theta"][time]
        psi = self.iekf_states["psi"][time]

        # Think this is the correct way to do it
        # TODO: is this the correct way to do it?
        r1 = R.from_rotvec([phi, 0, 0])
        r2 = R.from_rotvec([0, theta, 0])
        r3 = R.from_rotvec([0, 0, psi])
        rot_mat = r3.as_matrix() @ r2.as_matrix() @ r1.as_matrix()

        p = gtsam.Point3(x, y, z)
        v = gtsam.Point3(u, v, r)

        pose = Pose3(Rot3(rot_mat), p)
        state = NavState(pose, v)
        return state

    def depth_error(
        self,
        measurement: np.ndarray,
        this: gtsam.CustomFactor,
        values: gtsam.Values,
        jacobians: Optional[List[np.ndarray]],
    ) -> float:
        """
        Calculate the error betwen the depth sensor and the depth prediction
        """
        # Measurement is >0, meaning that it should be inverted
        assert measurement <= 0, "Measurement should be negative"
        key = this.keys()[0]
        estimate = values.atPose3(key)
        depth = estimate.z()
        error = measurement - depth
        if jacobians is not None:
            val = np.zeros((1, 6))
            val[0, -1] = 1
            jacobians[0] = val
        return error

    def velocity_error(
        self,
        measurement: np.ndarray,
        this: gtsam.CustomFactor,
        values: gtsam.Values,
        jacobians: Optional[List[np.ndarray]],
    ) -> float:
        """
        Calculate the error betwen the velocity prediction and the velocity measurement
        """
        key = this.keys()[0]
        vel_estimate = values.atPoint3(V(key))
        pose_estimate = values.atPose3(X(key))

        rot_mat = pose_estimate.rotation().matrix()
        vx = vel_estimate[0]
        vy = vel_estimate[1]
        vz = vel_estimate[2]
        v = np.array([vx, vy, vz]).reshape((3, 1))

        meas_t = measurement.T
        meas_world = rot_mat @ meas_t

        error = np.array(
            [
                meas_world[0, 0] - v[0, 0],
                meas_world[1, 0] - v[1, 0],
                meas_world[2, 0] - v[2, 0],
            ]
        )
        if jacobians is not None:
            jacobians[0] = rot_mat
        return error

    def plot_trajectories(self, result):
        """
        Compare the trajectories
        """
        res_poses = np.zeros((result.size() // 3, 6))

        for i in range(result.size() // 3):
            pose = result.atPose3(X(i))
            res_poses[i, 0] = pose.x()
            res_poses[i, 1] = pose.y()
            res_poses[i, 2] = pose.z()
            res_poses[i, 3] = pose.rotation().yaw()
            res_poses[i, 4] = pose.rotation().pitch()
            res_poses[i, 5] = pose.rotation().roll()

        # Create 3d plot
        fig = plt.figure()
        ax = fig.add_subplot(111, projection="3d")
        ax.plot(
            self.iekf_states["x"],
            self.iekf_states["y"],
            self.iekf_states["z"],
            label="IEKF",
        )
        ax.plot(res_poses[:, 0], res_poses[:, 1], res_poses[:, 2], label="GTSAM")
        ax.legend()
        plt.savefig(f"{sys.path[0]}/../images/isam/marginals.png")
        plt.show()

    def iSAM(self):
        """
        Optimize over the graph after each new observation is taken

        TODO: Test with only a few states at first
        """

        # Initialize a bunch of things
        isam = gtsam.ISAM2()
        graph = gtsam.NonlinearFactorGraph()
        initial = gtsam.Values()

        # Initialize the indices of the sensors to keep track of
        state_idx = 0
        # depth_idx = 0
        imu_idx = 0
        bias_idx = 0
        # vel_idx = 0
        node_idx = 0

        # Keep track of the time
        time_elapsed = 0
        total_time_elapsed = 0

        # Define how many steps to take
        state_step = self.iekf_times.shape[0]
        initial_time = self.iekf_times[0]

        while state_idx < state_step:
            state = self.get_nav_state(state_idx)
            state_time = self.iekf_times[state_idx] * 1e-18  # Convert to seconds

            if state_idx == 0:
                # Add prior to the graph at time 0
                priorPoseFactor = gtsam.PriorFactorPose3(
                    X(node_idx), state.pose(), self.priorNoise
                )
                graph.add(priorPoseFactor)

                priorVelFactor = gtsam.PriorFactorPoint3(
                    V(node_idx), state.velocity(), self.velNoise
                )
                graph.add(priorVelFactor)

                bx = self.imu["bx"][0]
                by = self.imu["by"][0]
                bz = self.imu["bz"][0]

                acc_bias = np.array([bx, by, bz])
                gyro_bias = np.array([bx, by, bz])
                initial_bias = gtsam.imuBias.ConstantBias(acc_bias, gyro_bias)

                graph.addPriorConstantBias(B(bias_idx), initial_bias, self.bias_cov)

                # Add values
                initial.insert(X(node_idx), state.pose())
                initial.insert(V(node_idx), state.velocity())
                initial.insert(B(bias_idx), initial_bias)

                node_idx += 1
            else:
                # Add bias factor periodically
                imu_time = self.imu_times[imu_idx] * 1e-18  # Convert to seconds
                while imu_time < state_time:

                    if imu_idx > 0:
                        prev_imu_time = (
                            self.imu_times[imu_idx - 1] * 1e-18
                        )  # Convert to seconds

                        imu_dt = (imu_time - prev_imu_time) * 1e9

                        # Get measured omega and a
                        omega_x = self.imu["omega_x"][imu_idx]
                        omega_y = self.imu["omega_y"][imu_idx]
                        omega_z = self.imu["omega_z"][imu_idx]

                        a_x = self.imu["ax"][imu_idx]
                        a_y = self.imu["ay"][imu_idx]
                        a_z = self.imu["az"][imu_idx]

                        measuredOmega = np.array([omega_x, omega_y, omega_z]).reshape(
                            (3, 1)
                        )
                        measuredAcc = np.array([a_x, a_y, a_z]).reshape((3, 1))
                        self.pim.integrateMeasurement(
                            measuredOmega, measuredAcc, imu_dt
                        )
                        print(f'IMU: {self.pim}')
                    imu_idx += 1
                    imu_time = self.imu_times[imu_idx] * 1e-18

            dt = self.iekf_times[state_idx] - self.iekf_times[state_idx - 1]
            if dt <= 0:
                # Skip this state, means that the udate step is made in this step
                state_idx += 1
                continue
            else:
                self.dt = dt * 1e-9
                time_elapsed += self.dt
                total_time_elapsed += self.dt

            # Check how long its been
            if time_elapsed > 0.5:
                print(f"####### New factors at time {total_time_elapsed} #######")
                imu_factor = gtsam.ImuFactor(
                    X(node_idx - 1),
                    V(node_idx - 1),
                    X(node_idx),
                    V(node_idx),
                    B(bias_idx),
                    self.pim,
                )
                graph.add(imu_factor)

                bias_idx += 1
                bias_factor = gtsam.BetweenFactorConstantBias(
                    B(bias_idx - 1),
                    B(bias_idx),
                    gtsam.imuBias.ConstantBias(),
                    self.bias_cov,
                )
                graph.add(bias_factor)

                depth_time = (self.depth_times - initial_time) * 1e-9
                depth_idx = np.argmin(np.abs(depth_time - total_time_elapsed))
                depth_time = depth_time[depth_idx]
                # print(depth_time, total_time_elapsed)
                depth_measurement = self.depth[depth_idx] * -1

                # Compute difference between current time and depth time
                depth_diff = abs(total_time_elapsed - depth_time)
                # print(depth_diff)

                if depth_diff < 1e-2:
                    # print(
                    # f'Below threshold for depth at time: {total_time_elapsed}!')
                    depth_factor = gtsam.CustomFactor(
                        self.depth_model,
                        [X(node_idx)],
                        partial(self.depth_error, np.array([depth_measurement])),
                    )
                    graph.add(depth_factor)

                initial.insert(B(bias_idx), gtsam.imuBias.ConstantBias())
                initial.insert(X(node_idx), state.pose())
                initial.insert(V(node_idx), state.velocity())

                self.pim.resetIntegration()

                node_idx += 1
                time_elapsed = 0

            isam.update(graph, initial)
            result = isam.calculateEstimate()
            graph.saveGraph(f"{sys.path[0]}/graph.dot", result)
            graph = gtsam.NonlinearFactorGraph()
            initial.clear()

            state_idx += 1
        return result, node_idx


def main():
    AUV_SLAM = AUViSAM()
    res, node_idx = AUV_SLAM.iSAM()
    print(res.size(), node_idx)
    AUV_SLAM.plot_trajectories(res)


if __name__ == "__main__":
    main()

File: src/cirs_girona_cala_viuda/scripts/frontend/front_end_isam.m
clear; close all; clc;
format long;
import gtsam.*;
fileIEKF = 'E:\US\Umich_Study\ROB530_MR\Project\UW_SLAM-master\UW_SLAM-master\src\cirs_girona_cala_viuda\data\states.csv';
fileIMU = 'E:\US\Umich_Study\ROB530_MR\Project\UW_SLAM-master\UW_SLAM-master\src\cirs_girona_cala_viuda\data\full_dataset\imu_adis_ros.csv';
fileStTime = 'E:\US\Umich_Study\ROB530_MR\Project\UW_SLAM-master\UW_SLAM-master\src\cirs_girona_cala_viuda\data\state_times.csv';
noiseModelPosr = noiseModel.Diagonal.Precisions(0.1*[1;1;1;1;1;1]);
firstPose = 2;
StatesSkip = 1; % Skip this many States measurements each time

%% Get initial conditions for the estimated trajectory
[states,state_times] = read_iekf_states(fileIEKF,fileStTime);
first_p = gtsam.Point3(states.x(firstPose), states.y(firstPose), states.z(firstPose));
first_rot = gtsam.Rot3.Ypr(states.psi(firstPose),states.theta(firstPose), states.phi(firstPose));
states.Position = [];
for i = 1:numel(state_times)
    states.Position{i} = gtsam.Point3(states.x(i), states.y(i), states.z(i));
end
%currentVelocityGlobal = LieVector([states.u(firstPose);states.v(firstPose);states.r(firstPose)]);
currentVelocityGlobal = LieVector([0;0;0]);
% currentVelocityGlobal = gtsam.Point3(states.u(firstPose), states.v(firstPose), states.r(firstPose)); % the vehicle is stationary at the beginning
currentPoseGlobal = gtsam.Pose3(first_rot, first_p); % initial pose is the reference frame (navigation frame)
currentBias = imuBias.ConstantBias(zeros(3,1), zeros(3,1));
sigma_init_x = noiseModel.Isotropic.Precisions([ 0.0; 0.0; 0.0; 1; 1; 1 ]);
sigma_init_v = noiseModel.Isotropic.Sigma(3, 1000.0);
sigma_init_b = noiseModel.Isotropic.Sigmas([ 0.100; 0.100; 0.100; 5.00e-05; 5.00e-05; 5.00e-05 ]);
acc_bias = [0.067; 0.115; 0.320];
gyro_bias = [0.067; 0.115; 0.320];
AccelerometerBiasSigma =  1.6700e-04;
GyroscopeBiasSigma = 2.9100e-06;
sigma_between_b = [AccelerometerBiasSigma * ones(3,1); GyroscopeBiasSigma * ones(3,1) ]; % BiasSigma
accSigma = 0.0100;
gyroSigma = 1.7500e-04;
integrationSigma = 0;
g = [0;0;-9.8];
w_coriolis = [0;0;0];
currentBias = gtsam.imuBias.ConstantBias(acc_bias, gyro_bias);
%% Solver object
isamParams = ISAM2Params;
isamParams.setFactorization('CHOLESKY');
isamParams.setRelinearizeSkip(10);
isam = gtsam.ISAM2(isamParams);
newFactors = NonlinearFactorGraph;
newValues = Values;

%% Main loop:
% (1) we read the measurements
% (2) we create the corresponding factors in the graph
% (3) we solve the graph to obtain and optimal estimate of robot trajectory
IMU = read_imu(fileIMU);

disp('-- Starting main loop: inference is performed at each time step, but we plot trajectory every 10 steps')

for measurementIndex = firstPose:size(states.x,2)
    
    % At each non=IMU measurement we initialize a new node in the graph
    currentPoseKey = symbol('x',measurementIndex);
    currentVelKey =  symbol('v',measurementIndex);
    currentBiasKey = symbol('b',measurementIndex);
    t = state_times(measurementIndex);
    
    if measurementIndex == firstPose
        %% Create initial estimate and prior on initial pose, velocity, and biases
        newValues.insert(currentPoseKey, currentPoseGlobal);
        newValues.insert(currentVelKey, currentVelocityGlobal);
        newValues.insert(currentBiasKey, currentBias);
        newFactors.add(PriorFactorPose3(currentPoseKey, currentPoseGlobal, sigma_init_x));
        newFactors.add(PriorFactorLieVector(currentVelKey, currentVelocityGlobal, sigma_init_v));
        newFactors.add(PriorFactorConstantBias(currentBiasKey, currentBias, sigma_init_b));
    else
        t_previous = state_times(measurementIndex-1);
        %% Summarize IMU data between the previous IEKF measurement and now
        IMUindices = find(IMU.times >= t_previous & IMU.times <= t);
        
        currentSummarizedMeasurement = gtsam.ImuFactorPreintegratedMeasurements( ...
            currentBias, accSigma.^2 * eye(3), ...
            gyroSigma.^2 * eye(3), integrationSigma.^2 * eye(3));
        
        if ~isempty(IMUindices)
            for imuIndex = IMUindices
                accMeas = [ IMU.ax(imuIndex); IMU.ay(imuIndex); IMU.az(imuIndex) ];
                omegaMeas = [ IMU.omegaX(imuIndex); IMU.omegaY(imuIndex); IMU.omegaZ(imuIndex)];
                deltaT = IMU.dt(imuIndex);
                currentSummarizedMeasurement.integrateMeasurement(accMeas, omegaMeas, deltaT);
            end
        else
            d=1
%             accMeas = [ IMU.ax(measurementIndex); IMU.ay(measurementIndex); IMU.az(measurementIndex) ];
%             omegaMeas = [ IMU.omegaX(measurementIndex); IMU.omegaY(measurementIndex); IMU.omegaZ(measurementIndex)];
%             deltaT = IMU.dt(measurementIndex);
%             currentSummarizedMeasurement.integrateMeasurement(accMeas, omegaMeas, deltaT);
        end


    % Create IMU factor
    newFactors.add(ImuFactor( ...
      currentPoseKey-1, currentVelKey-1, ...
      currentPoseKey, currentVelKey, ...
      currentBiasKey, currentSummarizedMeasurement, g, w_coriolis));
        
        % Bias evolution as given in the IMU data
        newFactors.add(BetweenFactorConstantBias(currentBiasKey-1, currentBiasKey, imuBias.ConstantBias(zeros(3,1), zeros(3,1)), ...
            noiseModel.Diagonal.Sigmas(sqrt(numel(IMUindices)) * sigma_between_b)));
        
        % Create state factor
        statesPose = Pose3(currentPoseGlobal.rotation, states.Position{measurementIndex});
        if mod(measurementIndex, StatesSkip) == 0
            newFactors.add(PriorFactorPose3(currentPoseKey, statesPose, noiseModelPosr));
        end
        
        % Add initial value
        newValues.insert(currentPoseKey, statesPose);
        newValues.insert(currentVelKey, currentVelocityGlobal);
        newValues.insert(currentBiasKey, currentBias);
        
        % Update solver
        % =======================================================================
        % We accumulate 2*states skip states before updating the solver at
        % first so that the heading becomes observable.
        if measurementIndex > firstPose + 2*StatesSkip
            isam.update(newFactors, newValues);
            newFactors = NonlinearFactorGraph;
            newValues = Values;
            
            if rem(measurementIndex,10)==0 % plot every 10 time steps
                cla;
                plot3DTrajectory(isam.calculateEstimate, 'g-');
                title('Estimated trajectory using ISAM2 (IMU)')
                xlabel('[m]')
                ylabel('[m]')
                zlabel('[m]')
                axis equal
                drawnow;
            end
            % =======================================================================
            currentPoseGlobal = isam.calculateEstimate(currentPoseKey);
            currentVelocityGlobal = isam.calculateEstimate(currentVelKey);
            currentBias = isam.calculateEstimate(currentBiasKey);
        end
    end
    
end % end main loop

%% Read IMU data
function [IMU] = read_imu(fileIMU)
disp('-- Reading IMU data')
IMU_Data = readcell(fileIMU);      % Read Data
IMU.times = []; IMU.dt = IMU_Data{2,1}; IMU.qx = []; IMU.qy = []; IMU.qz = []; IMU.qw = [];
IMU.ax = []; IMU.ay = []; IMU.az = []; IMU.omegaX = []; IMU.omegaY = []; IMU.omegaZ = [];
for row = 2:size(IMU_Data,1)
    %Time
    timeCol = 1;
    IMU.times = [IMU.times IMU_Data{row,timeCol}];
    if (row+1) <= size(IMU_Data,1)
        IMU.dt = [IMU.dt (IMU_Data{row+1,timeCol} - IMU_Data{row,timeCol})];
    end
    % Orientation
    qxCol = 4; qyCol = 5; qzCol = 6; qwCol = 7;
    IMU.qx = [IMU.qx IMU_Data{row,qxCol}]; IMU.qy = [IMU.qy IMU_Data{row,qyCol}]; IMU.qz = [IMU.qz IMU_Data{row,qzCol}]; IMU.qw = [IMU.qw IMU_Data{row,qwCol}];
    % Linear Acceleration
    axCol = 29; ayCol = 30; azCol = 31;
    IMU.ax = [IMU.ax IMU_Data{row,axCol}]; IMU.ay = [IMU.ay IMU_Data{row,ayCol}]; IMU.az = [IMU.az IMU_Data{row,azCol}];
    % Angular Velocity
    omegaxCol = 17; omegayCol = 18; omegazCol = 19;
    IMU.omegaX = [IMU.omegaX IMU_Data{row,omegaxCol}]; IMU.omegaY = [IMU.omegaY IMU_Data{row,omegayCol}]; IMU.omegaZ = [IMU.omegaZ IMU_Data{row,omegazCol}];
    % Acceleration Covariance
    IMU.accCov{row-1} = [IMU_Data{row,32}, IMU_Data{row,33}, IMU_Data{row,34}; IMU_Data{row,35}, IMU_Data{row,36}, IMU_Data{row,37}; IMU_Data{row,38}, IMU_Data{row,39}, IMU_Data{row,40}];
    % Omega Covariance
    IMU.omegaCov{row-1} = [IMU_Data{row,20}, IMU_Data{row,21}, IMU_Data{row,22}; IMU_Data{row,23}, IMU_Data{row,24}, IMU_Data{row,25}; IMU_Data{row,26}, IMU_Data{row,27}, IMU_Data{row,28}];
end
IMU.times = IMU.times.*1e-9;
IMU.dt = IMU.dt.*1e-9;
end

%% Read IEKF States

function [states,state_times] = read_iekf_states(fileIEKF,fileStTime)
disp('-- Reading IEKF data')
IEKF_Data = readcell(fileIEKF);      % Read Data
IEKF_time = readcell(fileStTime);      % Read Data
states.x = []; states.y = []; states.z = []; states.u = []; states.v = []; states.r = [];
states.phi = []; states.theta = []; states.psi = [];
state_times = [];
for row = 2:size(IEKF_Data,1)
    states.x = [states.x IEKF_Data{row,7}]; states.y = [states.y IEKF_Data{row,8}]; states.z = [states.z IEKF_Data{row,9}];   %p_x, p_y, p_z
    states.u = [states.u IEKF_Data{row,4}]; states.v = [states.v IEKF_Data{row,5}]; states.r = [states.r IEKF_Data{row,6}];   %v_x, v_y, v_z
    states.phi = [states.phi IEKF_Data{row,1}]; states.theta = [states.theta IEKF_Data{row,2}]; states.psi = [states.psi IEKF_Data{row,3}];   %theta_x, theta_y, theta_z
    state_times = [state_times  IEKF_time{row,1}];
end
state_times = state_times.*1e-9;
end
